In [1]:
import numpy as np
import pandas as pd
from scipy.stats import poisson
from warnings import filterwarnings
filterwarnings('ignore')


## 1. The number of customer returns in a retail chain per day follows a Poisson distribution at a rate of 25 returns per day. Write Python code to answer the following questions:

## (a) Calculate the probability that the number of returns exceeds 30 in a day.

In [2]:
# P(X=k) = (mu)k * e– (mu)/ k      #k= number of successes
1-poisson.cdf(k=30, mu=25)   #Probability Greater than Some Value
                            #mu= rate of change of event being happening
                            #e = a constant equal to approximately 2.71828

0.1366911308473363

## (b) If the chance of fraudulent return is 0.05, calculate the probability that there will be at least 2 fraudulent returns in any given day.

In [3]:
(1-poisson.cdf(k=2, mu=25))*0.05 

0.049999999764946554

## 2. A student is applying for Masters course in 8 US Universities and believes that she has in each of the eight universities a constant and independent 0.42 probability of getting selected. Write code to answer the following questions:

## (a) What is the probability that she will get call from at least 3 universities? 

In [4]:
from scipy import stats
1 - stats.binom.cdf(3, 8, 0.42)


0.45265884031065606

## (b) What is the probability that she will get calls from exactly 4 universities?

In [5]:
stats.binom.pmf(4, 8, 0.42)   #probability of exact number

0.2464944673605121

## 3. The time-of-failure of a machine follows exponential distribution with mean time between failures(MTBF) estimated to be 85 hrs. Write code to answer the following questions:

## (a) Calculate the probability that the system will fail before 85 hrs. 

In [6]:
stats.expon.cdf(x=85,loc=1/85, scale=85)


0.6320696377349755

## (b) Calculate the probability that it will not fail up to 150 hrs. 


In [7]:
1 - stats.expon.cdf(x=150, loc=1/85, scale=85)  #substracting from 1 because we are finding probability that machine will not fail.

0.17126084522766882

## 4. As per a survey on use of pesticides among 1000 farmers in grape farming for around 10 acres of grape farmland, it was found that the grape farmers spray 38 liters of pesticides in a week on an average with the corresponding standard deviation of 5 liters. Assume that the pesticide spray per week follows a normal distribution. Write code to answer the following questions:


## (a) What proportion of the farmers is spraying more than 50 liters of pesticide in a week? 

In [8]:
1 - stats.norm.cdf(50, 38, 5)

0.008197535924596155

## (b) What proportion of farmers is spraying less than 10 liters?

In [9]:
stats.norm.cdf(9, 38, 5)


3.3157459783261613e-09

## (c) What proportion of farmers is spraying between 30 liters and 60 liters? 


In [10]:
pro_60_liter = stats.norm.cdf(60, 38, 5) 
pro_60_liter


0.9999945874560923

In [11]:
pro_30_liter = stats.norm.cdf(30, 38, 5)
pro_30_liter

0.054799291699557974

In [12]:
pro_bet_30_and_60 = pro_60_liter-pro_30_liter
pro_bet_30_and_60

0.9451952957565343

## 5. A bottle filling machine fills water into 5 liters (5000 cm3) bottles. The company wants to test the null hypothesis that the average amount of water filled by the machine into the bottle is at least5000 cm3

* A random sample of 60 bottles coming out of the machine was selected and the exact contents of the selected bottles are recorded. The sample mean was 4,998.1 cm3
* The population standard deviation is known from the experience to be 1.30 cm3
* Assume that the population is normally distributed with the standard deviation of 1.30 cm3
* Write code to test the hypothesis at a of 5%. Explain the results.

In [13]:
import math

* Assume Null Hypothesis (H0) that the average amount of water filled by the machine into the bottle is at least 5000 cm3
* Alternate Hypothesis (H1) as the average amount of water filled by the machine into the bottle is at NOT least 5000 cm3

In [14]:
#making ztest as functions

def z_test(sample_mean,population_mean,std_deviation,sample_length):
    z_score=(sample_mean-population_mean)/(std_deviation/math.sqrt(sample_length))
    return z_score


In [15]:
z_test(4998.1,5000,1.30,60)

-11.321028242757974

In [16]:
z_critical_value = stats.norm.isf(q = 0.05)
z_critical_value

1.6448536269514729

* since  z score < critical value(1.645)
* so we accept the null hypothesis. In other words,average amount of water filled by the machine into the bottle is at least 5000 cm3

## 6. A fabric manufacturer would like to understand the proportion of defective fabrics they produce.His shop floor staffs have been stating that the percentage of defective is not more than 18%. He would like to test whether the claim made by his shop floor staff is correct. He picked up a random sample of 100 fabrics and found 22 defectives. Use α = 0.05 and write code to test the hypothesis that the percentage of defective components is less than 18%

In [18]:
!pip install scipy
from scipy.stats import binom_test
#sample_size = 100
#observed_defective = 22
#expected_defective <= 18

* Assume Null Hypothesis(H0) that the  percentage of defective is more than 18%
* Alternate Hypothesis (H1) that  percentage of defective is not more than 18% 

In [19]:
#calculating Binominal test which should be only 2 output
alpha = 0.05   #significance value
p_value = binom_test(x=22, n=100, p=0.18)  #calculating P value
p_value

0.29785331232911655

* Since P_value > significance value (alpha),so we will accept the null hypothesis. 
* In other words percentage of defective components is more than 18%


## 7.Suppose that the makers of ABC batteries want to demonstrate that their battery lasts an average of at least 60 min longer than their competitor brand. Two independent random samples of 100 batteries of each kind are selected from both the brands, and the batteries are used continuously.The sample average life of ABC is found to be 450 min. The average life of competitor batteries is 368 min with the sample standard deviation of 82 min and 78 min, respectively. Frame a hypothesis and write the code to test ABC’s claim at 95% significance.

In [20]:
#sample_size_1 = 100, sample_size_2 = 100
#mean1 = 450, mean2 = 368
#std1 = 82, std2 = 78


* Assume null hypothesis (HO) that the battery will lasts an average of at least 60 min longer than their competitor brand
* Alternate Hypothesis (H1) be that the battery will not lasts an average of at least 60 min longer than their competitor brand

In [21]:
stats.ttest_ind_from_stats(mean1=450,std1=82,nobs1=100,mean2=368,std2=78,nobs2=100)

Ttest_indResult(statistic=7.245580616896932, pvalue=9.390871476154758e-12)

* Since p value < alpha (0.05) ,we reject the null hypothesis and accept the alternate hypothesis. In other words,the battery will not lasts an average of at least 60 min longer than their competitor brand.